VIP-10: Storage node registration (on-chain)

tahpot opened the pull request.

Commit:
Initial commit

Commit:
Merge branch 'develop' into feature/vip-10

Commit:
Fix typo

Commit:
Add first draft

tahpot commented on VIPs/vip-10.md:
```diff
@@ -0,0 +1,109 @@
+
+# Specification
+
+## Data structures
```
@ITStar10 Can you propose the appropriate data structures

tahpot commented on VIPs/vip-10.md:
```diff
@@ -0,0 +1,109 @@
+
+## State variables
+
+TBA
```
@ITStar10 Can you propose appropriate state variables

ITStar10 replied:
- `countryCode` and `regionCode` are duplicated between StorageNode & Datacenter. Or are tehy different values?

ITStar10 replied:
This is the draft of state variables:
```
struct StorageNode {
  address didAddress;
  string endpointUri;
  uint datacenterId;

  uint unregisterTime;
}

struct Datacenter {
  string countryCode;
  string regionCode;
  uint lat;
  uint long;
}

/** mapping of storageNode ID => Storage*/
map(uint => StorageNode) private _storageNodeMap;
/** mapping of did address => storageNode ID*/
map(address => uint) private _addressNodeMap;
/** mapping of endpoint uri => storageNode ID*/
map(string => uint) private _endpointNodeMap;
/** mapping of countryCode => StorgeNode ID list*/
map(string => EnumerableSet.UintSet) private _countryNodes;
/** mapping of regionCode => StorageNode ID list*/
map(string => EnumerableSet.UintSet) private _regionNodes;

/** mapping of datacenterId => DataCenter */
map(uint => Datacenter) dataCenterMap
/** mapping of datacenterID => StorageNode ID list*/
map(uint => EnumerableSet.UintSet) private _dataCenterToNodeIdsmap;


CountersUpgradeable.Counter private _nodeIdCounter;
CountersUpgradeable.Counter private _dataCenterIdCounter;
```

tahpot replied:
>countryCode and regionCode are duplicated between StorageNode & Datacenter. Or are tehy different values?

Correct, duplicated.

If you have them in your mapping it's okay to only store them in `Datacenter`.

tahpot replied:
@ITStar10 Can you push a commit to add your proposed state variables?

ITStar10 replied:
It might be better to push the State Variables after contract development done.

tahpot replied:
@ITStar10 This is a draft, so put what you have in here for now and we will update at the end.

ITStar10 commented on VIPs/vip-10.md:
```diff
@@ -0,0 +1,109 @@
+
+An `establishmentDate` will be saved that matches the current unix timestamp (`block.timestamp`).
+
+## removeNode(didAddress: string, unregisterDatetime: uint, requestSignature: string)
```
I think, `unregisteredDatetime` is not necessary because we can get time in the blockchain itself.

ITStar10 commented:
@tahpot How can we name this contract?<br/>
Here are my suggestions:
- contract name: `StorageNodeContract` - package name:  `@verida/storage-node-contract`
- contract name: `VeridaNodeContract` - package name:  `@verida/node-contract`

tahpot commented:
@ITStar10 

I would like:

- Contract name: `StorageNodeRegistry`
- Package name: `@verida/storage-node-client`

ITStar10 commented:
@tahpot `@verida/storage-node-client` is for verida-js?

ITStar10 commented:
`countryCode` and `regionCode` in the StorageNode & Datacenter are the same?

tahpot commented:
> `countryCode` and `regionCode` in the StorageNode & Datacenter are the same?

No, different.

ITStar10 commented on VIPs/vip-10.md:
```diff
@@ -0,0 +1,109 @@
+
+Storage nodes must register themselves with the protocol so they can be discovered by new users. Storage nodes provide additional metadata (via `/status` endpoint) that assists users select the most appropriate storage nodes to use.
+
+It's not possible to re-use a DID to register multiple storage nodes. This is because storage endpoints sign their responses using the DID private key to enhance security of the protocol.
```
- A `didAddress` can be used only once for `addNode()` function?
- `endpointUri` is also unique in the contract? There can't be any same `endpointUri` with different `countryCode` or `regionCode`?

tahpot replied:
1. Yes
2. Yes

aurelticot commented on VIPs/vip-10.md:
```diff
@@ -0,0 +1,109 @@
+2. `unregisterDatetime`: The unix timestamp of when the storage node should no longer be available for selection. Must be at least 28 days in the future.
+3. `requestSignature`: The request parameters signed by the `didAddress` private key. Will be verified by `VDA-Verification-Base` library.
+
+Note: This does not mean the node is no longer active and managing data for users on the network. It just means it is no longer discoverable for new connections.
```
What's the motivation behind the 28-day notice (unregisterDatetime) if the `removeNode` is "just" disabling the discoverability?

tahpot replied:
Good question. See latest commit where I've added more detail.

aurelticot replied:
But, as the note suggests the discoverability and the operation (being active and managing data) of the node are **not related**, the existing users of the node are not affected by the discoverability or not.

Having a notice period if the node is shutting down, yes, because users have to move their data.

But for chaining the discoverability, I don't see the point.
I'd argue, Storage Nodes may want to be non-discoverable instantly in some cases (technical issues, legal issues, ...) so they don't have more users coming in.

tahpot replied:
I'm not following.

Can you suggest a change to illustrate what you're thinking?

aurelticot replied:
@tahpot, this passage:

> 2. `unregisterDatetime`: [...] will be removed from the network. [...] ensure users have sufficient time to migrate away [...]. [...] expect the node to become unavailable and their data to be deleted from `unregisterDatetime` onwards.

and this passage:

> Note: This does not mean the node is no longer active and managing data for users on the network. It just means it is no longer discoverable for new connections.

are contradictory.

I personally lean towards the Note:

- Unregistering a node **only** disables its discoverability for new users. No notice period because it doesn't affects its users.
- Shutting down a node is a different process. It requires a notice period because users must move away.

tahpot replied:
Hmm... I see what you mean, although I didn't intend it to come across as contradictory.

See the updated wording.

We don't want to support the concept of removing discoverability, but the node continues to operate for users. That will break the economic model when users are required to pay for storage via tokens. This contract is the source of truth of who is paying who.

aurelticot commented on VIPs/vip-10.md:
```diff
@@ -0,0 +1,109 @@
+
+Remove a data center by `id`. Will only remove the data center if there are no storage nodes using that datacenter.
+
+## getDatacenters(ids: string[]): Datacenter[]
```
For Storage Nodes to register themselves, they will have to pass the Id of their datacenter.
This `getDatacenters(ids)` taking a list of ids as parameter, doesn't make it easy for them to find their exact datacenter?
Maybe a `getAllDatacenters()` or a `getDatacentersByCountry(countryCode)`

tahpot replied:
Good point.

Perhaps we add:

- `getDatacentersByCountry(countryCode: string): DataCenter[]`
- `getDatacentersByRegion(regionCode: string): DataCenter[]`

Thoughts @itstar?

ITStar10 replied:
Ok, will add these functions.

tahpot replied:
@ITStar10 Please add them to this draft spec document as well

aurelticot commented on VIPs/vip-10.md:
```diff
@@ -0,0 +1,109 @@
+2. `endpointURI`: The storage node endpoint. Must be only one `endpointURI` registered at any time.
+3. `countryCode`: Unique two-character string code (see [VIP-7](./vip-7.md))
+4. `regionCode`: Unique region string code (see [VIP-7](./vip-7.md))
+5. `datacenterId`: Unique datacenter identifier. Must match a datacenter previously created with `addDatacenter()`.
```
> datacenter previously created with `addDatacenter()`

Feels like `addDatacenter` must be called by the Storage Node prior to call `addNode`, but `addDatacenter` is `ownerOnly`.

Maybe reword as "must match an existing datacenter, available via `getDatacenters`"

tahpot replied:
No, this is intentional.

At this stage, only we should be able to add data centres to avoid duplications etc.

In the future there could be a more complex decentralized voting process to add that type of info.

aurelticot replied:
Yep, I get the `ownerOnly` for creating a datacenter.
I was pointing the "previously created" may suggest they (the storage nodes) have to create the datacenter, but they should not (they can't), they must use an existing one.

tahpot replied:
Got it. Adjusted wording.

aurelticot commented on VIPs/vip-10.md:
```diff
@@ -0,0 +1,109 @@
+
+An `establishmentDate` will be saved that matches the current unix timestamp (`block.timestamp`).
+
+## removeNode(didAddress: string, unregisterDatetime: uint, requestSignature: string)
```
If a Storage Node wants to temporarily disable its discoverability, and so call `removeNode`, could it be able to register itself later with the same `didAddress`? Or does it have to use a different `didAddress`?

What's the impact (for the Storage Node and associated data, ...) if a different `didAddress` is needed?

Particularly looking at [@ITStar10 comment](https://github.com/verida/VIPs/pull/4/files#r1173420202) with the `StorageNode struct` defining a `unregisterTime property`

tahpot replied:
>If a Storage Node wants to temporarily disable its discoverability, and so call removeNode, could it be able to register itself later with the same didAddress? Or does it have to use a different didAddress?

Provided there is no active node registered with that `didAddress`, it should be able to register again with the same address.

aurelticot replied:
Ok, good.

Then, I'm wondering about the `unregisterTime` in @ITStar10 comment. It made me think the record won't be deleted but "flagged" as unregistered/inactive

ITStar10 replied:
That was a draft of `StorageNode` struct. <br/>
I am going to remove the `unregisteredTime` in the struct and create an individual variable.

aurelticot replied:
> That was a draft of `StorageNode` struct. I am going to remove the `unregisteredTime` in the struct and create an individual variable.

Ok. So that means the did is removed from the map(s) when calling `removeNode` so it can be added again later with `addNode`, right?

I mean, we don't have to go in the technical details in the spec, but the functional requirement is that a Storage Node can unregister the node and later register it again with the same DID.

Commit:
Add more detail on the removal of storage nodes.

ITStar10 commented:
## Remove datacenter ID
We should remove datacenter only when there are no storage nodes depending on. <br/>
What happened in following case?<br/>
We created a datacenter. And we added a storage node using this datacenter. After a while, we removed storage node.<br/>
**Can we remove the datacenter in case**?

tahpot commented:
> @tahpot `@verida/storage-node-client` is for verida-js?

Yes

tahpot commented:
> ## Remove datacenter ID
> We should remove datacenter only when there are no storage nodes depending on. What happened in following case? We created a datacenter. And we added a storage node using this datacenter. After a while, we removed storage node. **Can we remove the datacenter in case**?

No, data centres shouldn't be automatically deleted.

Commit:
Improve wording relating to data centers when adding a node.

ITStar10 committed:
Added `removeNodeComplete()` function
